### PR TITLE
Fix unsoundness in PDU loop

### PR DESCRIPTION
**Pull request by jamwaffles**

### CODE REVIEW

File: src/pdu_loop/mod.rs [hunk 1]
```diff
@@ -76,25 +78,51 @@ impl<const MAX_FRAMES: usize, const MAX_PDU_DATA: usize> PduStorage<MAX_FRAMES,
     }
 
     /// Get a reference to this `PduStorage` with erased lifetimes.
-    pub const fn as_ref(&self) -> PduStorageRef {
+    pub const fn as_ref<'a>(&'a self) -> PduStorageRef<'a> {
+        let num_frames = unsafe { &*self.frames.get() }.len();
```
Comment (jamesmunns) [inline src/pdu_loop/mod.rs:82]: This is just `MAX_FRAMES`, yeah?

File: src/pdu_loop/mod.rs [hunk 2]
```diff
@@ -76,25 +78,51 @@ impl<const MAX_FRAMES: usize, const MAX_PDU_DATA: usize> PduStorage<MAX_FRAMES,
...
+
+        if idx > self.num_frames {
+            return Err(Error::Pdu(PduError::InvalidIndex(idx)));
+        }
+
+        let frame = unsafe {
+            &mut *self
+                .frames
+                .as_ptr()
+                .add(core::mem::size_of::<pdu_frame::Frame>() * idx)
```
Comment (jamesmunns) [inline src/pdu_loop/mod.rs:115]:
```suggestion
                .add(idx)
```

`add(i)` already increments `size_of::<T>() * i`.

File: src/pdu_loop/mod.rs [hunk 3]
```diff
@@ -76,25 +78,51 @@ impl<const MAX_FRAMES: usize, const MAX_PDU_DATA: usize> PduStorage<MAX_FRAMES,
...
+        };
+        let data = unsafe {
+            core::slice::from_raw_parts_mut(
+                self.frame_data.as_ptr().add(self.frame_data_len * idx),
```
Comment (jamesmunns) [inline src/pdu_loop/mod.rs:119]: This is right, because our `T` here is `u8`, not `[u8; frame_data_len]`

**Comment by jamwaffles** [comment]
I just pushed a commit to do an experiment with the `atomic_enum` crate, which gives me this MIRI error:

```
error: Undefined Behavior: Data race detected between Write on thread `TX task` and Write on thread `pdu_loop::tests::parallel` at alloc536+0x800
   --> src\pdu_loop\mod.rs:109:30
    |
109 |         let frame = unsafe { &mut *self.frames.as_ptr().add(idx) };
    |                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Data race detected between Write on thread `TX task` and Write on thread `pdu_loop::tests::parallel` at alloc536+0x800
    |
    = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
    = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
    = note: BACKTRACE:
    = note: inside `pdu_loop::PduStorageRef::<'_>::frame` at src\pdu_loop\mod.rs:109:30: 109:65
note: inside `pdu_loop::PduLoop::send_frames_blocking::<[closure@src\pdu_loop\mod.rs:406:64: 406:77]>`
   --> src\pdu_loop\mod.rs:167:33
    |
167 |             let (frame, data) = self.storage.frame(idx)?;
    |                                 ^^^^^^^^^^^^^^^^^^^^^^^
note: inside closure
   --> src\pdu_loop\mod.rs:405:25
    |
405 | /                         PDU_LOOP
406 | |                             .send_frames_blocking(ctx.waker(), |frame, data| {
407 | |                                 let packet = frame
408 | |                                     .write_ethernet_packet(&mut packet_buf, data)
...   |
418 | |                                 Ok(())
419 | |                             })
    | |______________________________^
```

The `FrameState` is still inside `PduFrame` so moving it out as suggested by you @jamesmunns may fix this - which I'll explore tomorrow - but for now it's erroring out. 

I'll leave it here for tonight 😪

**Comment by jamwaffles** [comment]
I forgot to note last night that all the ordering for the atomic operations is just a blanket "I dunno" `SeqCst` which is likely causing the data race issue picked up by MIRI.

Also added MIRI to the tests so output is more visible.

File: src/pdu_loop/frame_element.rs [hunk 4]
```diff
@@ -0,0 +1,494 @@
...
+
+// Used to store a FrameElement with erased const generics
+#[derive(Debug)]
+pub struct FrameBox<'sto> {
+    pub frame: NonNull<FrameElement<0>>,
+    pub _lifetime: PhantomData<&'sto mut FrameElement<0>>,
+}
+
+// FIXME: This seems bad. The examples break without it though
+unsafe impl<'sto> Send for FrameBox<'sto> {}
```
Comment (jamwaffles) [inline src/pdu_loop/frame_element.rs:152]: Note for review

File: src/pdu_loop/frame_element.rs [hunk 5]
```diff
@@ -0,0 +1,494 @@
...
+
+    unsafe fn take_waker(&self) -> Option<Waker> {
+        (*addr_of!((*self.frame.as_ptr()).frame.waker))
+            .try_write()
+            .expect("Contention take_waker")
+            .take()
+    }
+
+    // FIXME: Is interior mutability ok here? MIRI isn't complaining about race conditions so
+    // _maybe_ we're ok? The 3 fields aren't set anywhere else except during frame initialisation.
```
Comment (jamwaffles) [inline src/pdu_loop/frame_element.rs:170]: Note for review

File: src/pdu_loop/frame_element.rs [hunk 6]
```diff
@@ -0,0 +1,494 @@
...
+    }
+}
+
+pub struct RxFrameDataBuf<'sto> {
+    _frame: ReceivedFrame<'sto>,
+    data_start: NonNull<u8>,
+    data_end: NonNull<u8>,
+}
+
+// FIXME: This seems bad. The examples break without it though
```
Comment (jamwaffles) [inline src/pdu_loop/frame_element.rs:469]: Note for review

File: src/pdu_loop/frame_element.rs [hunk 7]
```diff
@@ -0,0 +1,494 @@
...
+
+impl<'sto> Drop for ReceivedFrame<'sto> {
+    fn drop(&mut self) {
+        log::trace!("Drop frame element");
+
+        unsafe { FrameElement::set_state(self.inner.frame, FrameState::None) }
+    }
+}
+
+pub struct RxFrameDataBuf<'sto> {
```
Comment (jamwaffles) [inline src/pdu_loop/frame_element.rs:463]: I'd like to understand how this solves lifetime issues in `send_coe_service`. I'm missing a lot of Rust-fu but also feel the answer is reasonably simple :D

File: src/pdu_loop/frame_element.rs [hunk 8]
```diff
@@ -0,0 +1,494 @@
...
+    /// The frame is being sent over the network interface.
+    Sending = 3,
+    /// A frame response has been received and is now ready for parsing.
+    RxBusy = 5,
+    /// Frame response parsing is complete. The frame and its data is ready to be returned in
+    /// `Poll::Ready`.
+    RxDone = 6,
+    /// The frame TX/RX is complete, but the frame is still in use by calling code.
+    RxProcessing = 7,
+}
```
Comment (jamwaffles) [inline src/pdu_loop/frame_element.rs:49]: Fewer variants than in https://github.com/jamesmunns/ethercrab/compare/master...jamesmunns:ethercrab:james-hacking as I wasn't sure I needed the granularity. `WAIT_RX` and `DONE` are removed.

File: src/pdu_loop/frame_element.rs [hunk 9]
```diff
@@ -0,0 +1,490 @@
...
+
+#[atomic_enum::atomic_enum]
+#[derive(PartialEq, Default)]
+pub enum FrameState {
+    // SAFETY: Because we create a bunch of `Frame`s with `MaybeUninit::zeroed`, the `None` state
+    // MUST be equal to zero. All other fields in `Frame` are overridden in `replace`, so there
+    // should be no UB there.
+    /// The frame is ready to be claimed
+    #[default]
+    None = 0,
```
Comment (jamesmunns) [inline src/pdu_loop/frame_element.rs:34]: Note: only create_frame is allowed to touch "empty frames", this is a "free allocation block".

File: src/pdu_loop/frame_element.rs [hunk 10]
```diff
@@ -0,0 +1,490 @@
...
+    /// The frame is claimed and can be initialised ready for sending.
+    Created = 1,
```
Comment (jamesmunns) [inline src/pdu_loop/frame_element.rs:36]: Buffer has been zeroed, frame has been given out by create_frame

File: src/pdu_loop/frame_element.rs (hunk 4 above)
Comment (jamwaffles) [reply src/pdu_loop/frame_element.rs:148]: This is ok.

But TODO: Only put this on typestates that need it.

File: src/pdu_loop/frame_element.rs [hunk 11]
```diff
@@ -0,0 +1,490 @@
...
+            .expect("Contention replace_waker")
+            .replace(waker);
+    }
+
+    unsafe fn take_waker(&self) -> Option<Waker> {
+        (*addr_of!((*self.frame.as_ptr()).frame.waker))
+            .try_write()
+            .expect("Contention take_waker")
+            .take()
+    }
```
Comment (jamwaffles) [inline src/pdu_loop/frame_element.rs:163]: TODO: Move to the right wrapping typestate

**Review by jamwaffles — COMMENTED** [review]
Some notes made from a call with James M

- State diagram
  - What invariants are true at each state
  - What transitions are allowable from each state
  - Who is allowed to make those transitions
  - What defences are there to ensure the transition only happens once (e.g. two people can't start the transition at the same time)
    - Exclusive ownership with typestates
    - Atomic compare/swap
  - Leading vs trailing invairant (e.g. claim, then init vs init then claim)
- Try using `loom` against just driving the states forward
- Try getting rid of spinlock - mutex on top of mutex is bad
  - TX/RX must be !Sync to guarantee waker can only be accessed from one thread

**Pull request merged**